src/compiler/emitter.ts

```diff
@@ -4520,16 +4520,18 @@ namespace ts {
                 return false;
             }
 
+            // If nodes haven't been transformed, they should always be real siblings
+            if (!previousNode.original && !nextNode.original) {

```

Comment:
`original` is only set when using a `factory.updateX` method or when calling `setOriginalNode`, so this isn't exactly the best test. As written, this only tests whether an `original` pointer has not been set, which is also true for any `factory.createX` method in any transform.

If you want to _only_ focus on parse tree nodes (i.e., nodes that were created by the parser), use `isParseTreeNode` or check for `!(node.flags & NodeFlags.Synthesized)`.

We have a bunch of conflicting terms around `Node` objects that I've been meaning to clarify. Functions like `nodeIsSynthesized` aren't *quite* accurately named, since they are checking for "non-existent positions" (i.e., `pos` or `end` is `-1`). The function really should have been something like `textRangeIsMissing` or something.

Terminology wise:
- A "Parse Tree Node" is a `Node` produced by the parser (`(node.flags & NodeFlags.Synthesized) === 0`).
  - A "Parse Tree Node" will *always* have a valid (i.e. `>= 0`) `pos` and `end` (though they may be the same, indicating a missing node).
  - A "Parse Tree Node" will have a `parent` pointer if:
    - The node is not a `SourceFile` (they don't have parents)
    - The containing `SourceFile` was either parsed with `true` for `setParentNodes`, or was bound.
- A "Synthetic Node" is a `Node` produced outside of the parser (`(node.flags & NodeFlags.Synthesized) !== 0`).
  - A "Synthetic Node" *may* have valid `pos` and `end` values for various reasons:
    - A "Synthetic Node" created from an `update` method will copy the `pos` and `end` of its original node.
    - A "Synthetic Node" may have its `pos` and `end` copied from a source node via `setTextRange`.
      - This is often the case for copies of an `Identifier` or literal that came from the parse tree as a means of preserving the quote style or escape characters used in the original source. The copies are often used when the node needs to be used twice with different source map or comment handling. 
    - A "Synthetic Node" can have a mix of valid (`>=0`) or invalid (`-1`) values for `pos` and `end`.
- An "Original Node" is a `Node` that was the source of a transformation (often through calling an `update` method), but an "Original Node" could be *either* a "Parse Tree Node" or a "Synthetic Node".
  - Some "Parse Tree Nodes" *will* have `original` pointers due to some cases in `Parser` where we reparse a node (such as with Top-Level Await). There will likely be more of these cases as at some point I'd like to introduce `ObjectAssignmentPattern` and `ArrayAssignmentPattern` as a reparse of `ObjectLiteralExpression` and `ArrayLiteralExpression` in assignment patterns (our current approach for handling those in destructuring is overly complex).

tldr: If the goal here is to only perform this check on adjacent "Parse Tree Nodes" that belong to the same parent, then you probably want to use `isParseTreeNode(previousNode) && isParseTreeNode(nextNode) && previousNode.parent === nextNode.parent`.